jsonschema/reference.go

```diff
@@ -0,0 +1,161 @@
+package jsonschema
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/franela/goreq"
+)
+
+// jsonRef represents a JSON Reference source and targetRef
+type jsonRef struct {
+	Source []string
+	Target string
+}
+
+// dereference parses JSON string and replaces all $ref with the referenced data.
+func dereference(schemaPath string, input []byte, isTop bool) ([]byte, error) {
+	if !strings.Contains(string(input), "$ref") {
+		return input, nil
+	}
+
+	var data interface{}
+	json.Unmarshal([]byte(input), &data)
+	refs, err := walkInterface(data, []string{}, []jsonRef{})
+	if err != nil {
+		return input, fmt.Errorf("unable to walk interface %s: %v", schemaPath, err)
+	}
+
+	for _, ref := range refs {
+		top := data
+		for i, item := range ref.Source {
+			if isTop && (ref.Source[0] == "allOf" || ref.Source[0] == "oneOf") {
+				continue // do not dereference top-level allOr and oneOf
```

Comment:
By doing this the former handling of allOf and oneOf doesn't change. That makes sense to do as an interim but I'm not sure it makes sense to do long term. Can you follow up this merging with a new ticket to investigate handling allOf/oneOf in the same way as all other refs? That would allow us to do things like export this function it could be usable by anyone importing the library as a standalone dereference utility.

Reply:
Agreed, I had thought of extending the allOf/oneOf in the jstransform logic, but even more complex than the scope of this ticket (allOf may be less complex than oneOf).  I will create a follow up ticket.

Reply:
Adding PENG-4337 to non-sprint, most likely to fall over into the next sprint.